kernel/memory/src/paging/mod.rs

```diff
@@ -317,18 +333,20 @@ pub fn init(
 
         // Handle the stack (a separate data section), which consists of one guard page followed by the real stack pages.
         // It does not need to be identity mapped because each AP core will have its own stack.
-        let stack_pages = page_allocator::allocate_pages_by_bytes_at(stack_start_virt, (stack_end_virt - stack_start_virt).value())?;
-        let start_of_stack_pages = *stack_pages.start() + 1; 
-        let (stack_guard_page, stack_allocated_pages) = stack_pages.split(start_of_stack_pages)
-            .map_err(|_ap| "BUG: initial stack's allocated pages were not split correctly after guard page")?;
-        let stack_start_frame = Frame::containing_address(stack_start_phys) + 1; // skip 1st frame, which corresponds to the guard page
-        let stack_allocated_frames = frame_allocator::allocate_frames_at(stack_start_frame.start_address(), stack_allocated_pages.size_in_pages())?;
-        let stack_mapped_pages = mapper.map_allocated_pages_to(
-            stack_allocated_pages,
-            stack_allocated_frames,
-            data_flags,
-        )?;
-        stack_page_group = Some((stack_guard_page, NoDrop::new(stack_mapped_pages)));
+        let stack_guard_page = page_allocator::allocate_pages_at(stack_start_virt, 1)?;
+        let mut stack_mapped_pages: Option<MappedPages> = None;
+        let mut iter = stack_mappings.iter();
+        while let Some(Some((page, frame))) = iter.next() {
+            let allocated_page = page_allocator::allocate_pages_at(page.start_address(), 1)?;
+            let allocated_frame = frame_allocator::allocate_frames_at(frame.start_address(), 1)?;
+            let mapped_pages = mapper.map_allocated_pages_to(allocated_page, allocated_frame, data_flags)?;
+            if let Some(ref mut stack_mapped_pages) = stack_mapped_pages {
+                stack_mapped_pages.merge(mapped_pages).map_err(|_| "failed to merge stack mapped pages")?;
+            } else {
+                stack_mapped_pages = Some(mapped_pages);
+            }
```

Comment:
this is a new and semi-interesting use case for `MappedPages`. 

Should I perhaps add a new API to `Mapper` or change the existing one to something like
```rust
    pub fn map_allocated_pages_to(
        &mut self,
        pages: AllocatedPages,
        frames: impl IntoIter<Item = AllocatedFrames>,
        flags: impl Into<PteFlagsArch>,
    ) -> Result<MappedPages, &'static str> { ... }
```
such that the caller can provide a series of `AllocatedFrames` objects that would be mapped by the `AllocatedPages` in sequential order of iteration

(not for this PR, just a general thought)

Reply:
Probably? I do think it's strange that we mostly only map contiguous pages to contiguous frames, despite one of the benefits of paging being that contiguous pages can be mapped to non-contiguous frames.

The frames should probably be an [`ExactSizeIterator`](https://doc.rust-lang.org/core/iter/trait.ExactSizeIterator.html) though so that we can ensure there are as many frames as pages prior to mapping anything. `ExactSizeIterator` is implemented for `array::IntoIter`.

Reply:
`ExactSizeIterator` would be nice to have, but I don't know if it would always be possible to have an array or vec (or whatever form of known-sized storage) in other use cases. But we could certainly specialize on that with a function that accepted an `ExactSizeIterator`.

Reply:
If the user doesn't know how many frames they have, how could they allocate the correct number of pages?